sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,753 @@
+provide the benefit of re-allocating memory from the memory pool.
+
+[source,c++]
+----
+int main(int argc, char *argv[])
+{
+  queue q(property::queue::in_order);
+
+  ext::oneapi::experimental::memory_pool memPool
+    = q.get_context().ext_oneapi_get_default_memory_pool(usm::alloc::device);
+
+  memPool.set_new_threshold(1024);
+  
+  {
+    // memory pool allocates memory from the SYCL runtime
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for first kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  // memory pool does not release memory back to the SYCL runtime as it is still
+  // within the specified threshold
+  q.wait();
+
+  {
+    // memory pool re-uses previously allocated memory
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for second kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something_else(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  // again memory pool does not release memory back to the SYCL runtime
+  q.wait();
+}
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Memory pool
+
+This extension introduces the `memory_pool` class, which provides a handle to a
+memory pool owned by the SYCL runtime or a specific backend, and adheres to the
+SYCL common reference semantics.
+
+Memory pools have the following properties:
+
+* A memory pool can allocate memory from two possible sources; either the SYCL
+  runtime or an existing USM memory allocation provided by the user. The default
+  source is the SYCL runtime.
+* A maximum allocation size (in bytes) is used to manage the total amount of
+  memory which can be allocated in the memory pool. If the maximum size is
+  exceeded an error is thrown. The default maximum size is
+  implementation-defined.
+* A deallocation threshold (in bytes) is used to determine how much memory the
+  SYCL runtime should aim to maintain in the memory pool, without releasing back
+  to the source. The default deallocation threshold is zero.
```

Comment:
`source` is an overloaded word. Can you clarify the wording here a bit more, please.

Reply:
I think in general I don't understand what this property is doing. `Used to determine`, `should`, ... is quite vague phrasing IMO. It would be good if you supplemented this with an example.